### PR TITLE
Emilhotkowski/added endpoints for rate limiting

### PR DESCRIPTION
**Pull request by emilhotkowski**
### Description

Added endpoints and domain model for RateLimiting.

### Backwards Compatible
 
Yes

### Tested

No tests yet.

### CODE REVIEW

File: packages/phone-number-privacy/signer/src/database/wrappers/domain.ts [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+function domains() {
+  return getDatabase()<Domain>(DOMAINS_TABLE)
+}
+
+export async function setDomainDisabled(
+  domain: string,
+  value: boolean,
+  logger: Logger
+): Promise<boolean> {
+  const storeRequestMeter = Histograms.dbOpsInstrumentation.labels('disableDomain').startTimer()
```
Comment (alecps) [inline packages/phone-number-privacy/signer/src/database/wrappers/domain.ts]: nit: disableDomainMeter

File: packages/phone-number-privacy/signer/src/server.ts [hunk 2]
```diff
@@ -17,8 +18,14 @@ export enum Endpoints {
   METRICS = '/metrics',
   GET_BLINDED_MESSAGE_PARTIAL_SIG = '/getBlindedMessagePartialSig',
   GET_QUOTA = '/getQuota',
+
+  DISABLE_DOMAIN = '/domain/disable/:domain',
```
Comment (alecps) [inline packages/phone-number-privacy/signer/src/server.ts]: Why did we add the ":domain" ?

File: packages/phone-number-privacy/signer/src/database/wrappers/domain.ts [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+  [DOMAINS_COLUMNS.domain]: '',
+  [DOMAINS_COLUMNS.timer]: 0,
+}
+
+export async function setDomainDisabled(
+  domain: string,
+  value: boolean,
+  logger: Logger
+): Promise<boolean> {
+  const disableDomainMeter = Histograms.dbOpsInstrumentation.labels('disableDomain').startTimer()
```
Comment (alecps) [inline packages/phone-number-privacy/signer/src/database/wrappers/domain.ts]: we should add 'disableDomain' to the labels in common/metrics

File: packages/phone-number-privacy/signer/src/database/wrappers/domain.ts [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+}
+
+const emptyDomain: Domain = {
+  [DOMAINS_COLUMNS.counter]: 0,
+  [DOMAINS_COLUMNS.disabled]: false,
+  [DOMAINS_COLUMNS.domain]: '',
+  [DOMAINS_COLUMNS.timer]: 0,
+}
+
+export async function setDomainDisabled(
```
Comment (alecps) [inline packages/phone-number-privacy/signer/src/database/wrappers/domain.ts]: Nit: Because disabling domains is permanent, we can probably just call this "disableDomain" and not accept 'value' as a parameter (we'll just always set it to true).

File: packages/phone-number-privacy/signer/src/database/wrappers/domain.ts [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+  domain: string,
+  value: boolean,
+  logger: Logger
+): Promise<boolean> {
+  const disableDomainMeter = Histograms.dbOpsInstrumentation.labels('disableDomain').startTimer()
+  logger.debug({ domain, value }, 'Disabling domain')
+  try {
+    await domains().where(DOMAINS_COLUMNS.domain, domain).update(DOMAINS_COLUMNS.disabled, value)
+    disableDomainMeter()
```
Comment (alecps) [inline packages/phone-number-privacy/signer/src/database/wrappers/domain.ts]: we should also add .timeout(DB_TIMEOUT) here

File: packages/phone-number-privacy/signer/src/database/wrappers/domain.ts [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+    Counters.databaseErrors.labels(Labels.update).inc()
+    logger.error(ErrorMessage.DATABASE_UPDATE_FAILURE)
+    logger.error(err)
+    disableDomainMeter()
+    throw err
+  }
+}
+
+export async function getDomain(domain: string, logger: Logger): Promise<Domain | null> {
+  const getDomainMeter = Histograms.dbOpsInstrumentation.labels('getDomain').startTimer()
```
Comment (alecps) [inline packages/phone-number-privacy/signer/src/database/wrappers/domain.ts]: Same as above. We need to add all these labels in /common/metrics

File: packages/phone-number-privacy/signer/src/database/wrappers/domain.ts [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+  logger.debug({ domain }, 'Get Domain')
```
Comment (alecps) [inline packages/phone-number-privacy/signer/src/database/wrappers/domain.ts]: Nit: consider something like 'Getting domain from db'

File: packages/phone-number-privacy/signer/src/database/wrappers/domain.ts [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+  } catch (err) {
+    Counters.databaseErrors.labels(Labels.update).inc()
+    logger.error(ErrorMessage.DATABASE_UPDATE_FAILURE)
+    logger.error(err)
+    disableDomainMeter()
+    throw err
+  }
+}
+
+export async function getDomain(domain: string, logger: Logger): Promise<Domain | null> {
```
Comment (alecps) [inline packages/phone-number-privacy/signer/src/database/wrappers/domain.ts]: This would never resolve to null, right?

File: packages/phone-number-privacy/signer/src/database/wrappers/domain.ts [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+  const getDomainMeter = Histograms.dbOpsInstrumentation.labels('getDomain').startTimer()
+  logger.debug({ domain }, 'Get Domain')
+  try {
+    const result = await domains().where(DOMAINS_COLUMNS.domain, domain).first().timeout(DB_TIMEOUT)
+    getDomainMeter()
+    return result ?? { ...emptyDomain }
```
Comment (alecps) [inline packages/phone-number-privacy/signer/src/database/wrappers/domain.ts]: Nit: I think we can just return emptyDomain here rather than copying it to a new object

File: packages/phone-number-privacy/signer/src/database/wrappers/domain.ts [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+  counter: number,
+  timer: number,
+  logger: Logger
+): Promise<void> {
+  const updateDomainMeter = Histograms.dbOpsInstrumentation.labels('updateDomain').startTimer()
+  logger.debug({ domain, timer, counter }, 'Update Domain')
+  try {
+    const result = await domains().where(DOMAINS_COLUMNS.domain, domain).first().timeout(DB_TIMEOUT)
+
+    if (result == null) {
```
Comment (alecps) [inline packages/phone-number-privacy/signer/src/database/wrappers/domain.ts]: Consider if(!result)

File: packages/phone-number-privacy/signer/src/database/models/domain.ts [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+  timer = 'timer',
+  disabled = 'disabled',
+}
+export class Domain {
+  [DOMAINS_COLUMNS.domain]: string;
+  [DOMAINS_COLUMNS.counter]: number;
+  [DOMAINS_COLUMNS.timer]: number;
+  [DOMAINS_COLUMNS.disabled]: boolean
+
+  constructor(domain: string, counter: number, timer: number, disabled: boolean) {
```
Comment (alecps) [inline packages/phone-number-privacy/signer/src/database/models/domain.ts]: We should probably hardcode disabled to false here, since we should never be constructing a disabled Domain

File: packages/phone-number-privacy/signer/src/database/wrappers/domain.ts (hunk 9 above)
Comment (alecps) [reply packages/phone-number-privacy/signer/src/database/wrappers/domain.ts]: On second thought, I think we should return null here if the record does not exist and resolve to default values in the handler. Then we can reuse this to check if the record exists in updateDomain and handleDisableDomain functions. Sorry for the back and forth

File: packages/phone-number-privacy/signer/src/domain/response/domainStatus.response.ts [hunk 12]
```diff
@@ -0,0 +1,6 @@
+export interface DomainStatusResponse {
```
Comment (alecps) [inline packages/phone-number-privacy/signer/src/domain/response/domainStatus.response.ts]: I think we may want to define this in the common package

File: packages/phone-number-privacy/signer/src/domain/domain.service.ts [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+    }
+  }
+
+  public async handleGetDomainStatus(
+    request: Request<core.ParamsDictionary, {}, {}>,
+    response: Response
+  ): Promise<void> {
+    const logger = response.locals.logger
+    logger.info('Getting domain status', { domain: request.params.domain })
+    const result = await getDomain(request.params.domain, logger)
```
Comment (alecps) [inline packages/phone-number-privacy/signer/src/domain/domain.service.ts]: We need to catch and handle the error for getDomain

File: packages/phone-number-privacy/signer/src/domain/domain.service.ts [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+    if (result) {
+      const resultResponse: DomainStatusResponse = {
+        domain: result[DOMAINS_COLUMNS.domain],
+        counter: result[DOMAINS_COLUMNS.counter],
+        disabled: result[DOMAINS_COLUMNS.disabled],
+        timer: result[DOMAINS_COLUMNS.timer],
+      }
+      response.status(200).send(resultResponse)
+    } else {
+      response.status(500).send({ message: 'Cannot get status' })
```
Comment (alecps) [inline packages/phone-number-privacy/signer/src/domain/domain.service.ts]: We should use the respondWithError helper function here

File: packages/phone-number-privacy/signer/src/domain/domain.service.ts [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+
+  public async handleDisableDomain(
+    request: Request<core.ParamsDictionary, {}, DisableDomainRequestBody>,
+    response: Response
+  ): Promise<void> {
+    const logger = response.locals.logger
+    logger.info('Disabling domain', { domain: request.params.domain })
+
+    const disabled = request.body.value ?? true
+    const result = await setDomainDisabled(request.params.domain, disabled, logger)
```
Comment (alecps) [inline packages/phone-number-privacy/signer/src/domain/domain.service.ts]: These db functions need error handling

File: packages/phone-number-privacy/signer/src/domain/domain.service.ts [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+
+    if (result) {
+      response.sendStatus(200)
+    } else {
+      response.status(500).send({ message: 'Cannot disable the domain' })
```
Comment (alecps) [inline packages/phone-number-privacy/signer/src/domain/domain.service.ts]: Please use respondWithError

File: packages/phone-number-privacy/common/src/interfaces/requests.ts [hunk 17]
```diff
@@ -20,4 +20,8 @@ export interface GetQuotaRequest {
   sessionID?: string
 }
 
+export interface DisableDomainRequestBody {
```
Comment (alecps) [inline packages/phone-number-privacy/common/src/interfaces/requests.ts]: This needs to also have a domain field

**Review by alecps — CHANGES_REQUESTED** [review]
Good work! Left a few comments

File: packages/phone-number-privacy/signer/src/database/wrappers/domain.ts (hunk 3 above)
Comment (emilhotkowski) [reply packages/phone-number-privacy/signer/src/database/wrappers/domain.ts]: Are we sure we have to do that? I've checked "_getPerformedQueryCount_" from account.ts file. And we never declare it anywhere.

File: packages/phone-number-privacy/signer/src/database/wrappers/domain.ts (hunk 6 above)
Comment (emilhotkowski) [reply packages/phone-number-privacy/signer/src/database/wrappers/domain.ts]: Same question, if we need to add these labels, where to do it?

File: packages/phone-number-privacy/signer/src/database/wrappers/domain.ts (hunk 5 above)
Comment (emilhotkowski) [reply packages/phone-number-privacy/signer/src/database/wrappers/domain.ts]: Added

File: packages/phone-number-privacy/signer/src/database/wrappers/domain.ts (hunk 8 above)
Comment (emilhotkowski) [reply packages/phone-number-privacy/signer/src/database/wrappers/domain.ts]: Changed from empty to null domain

**Comment by emilhotkowski** [comment]
@alecps what should happen if we don't find a domain on "GetDomainStatus" operation? Service right now answers with an error 500

File: packages/phone-number-privacy/signer/src/database/wrappers/domain.ts (hunk 3 above)
Comment (alecps) [reply packages/phone-number-privacy/signer/src/database/wrappers/domain.ts]: Ahh you're totally right. My bad

File: packages/phone-number-privacy/signer/src/domain/domain.service.ts [hunk 18]
```diff
@@ -1,49 +1,75 @@
...
+            Endpoints.DOMAIN_STATUS,
+            response,
+            500,
+            ErrorMessage.DATABASE_GET_FAILURE
+          )
+        }
+      })
+      .catch((e) => {
+        logger.error('Error while getting domain status', e)
+        respondWithError(Endpoints.DISABLE_DOMAIN, response, 500, ErrorMessage.DATABASE_GET_FAILURE)
```
Comment (alecps) [inline packages/phone-number-privacy/signer/src/domain/domain.service.ts]: should be Endpoints.DOMAIN_STATUS

File: packages/phone-number-privacy/signer/src/domain/domain.service.ts [hunk 19]
```diff
@@ -1,49 +1,75 @@
...
+        if (result) {
+          const resultResponse: DomainStatusResponse = {
+            domain: result[DOMAINS_COLUMNS.domain],
+            counter: result[DOMAINS_COLUMNS.counter],
+            disabled: result[DOMAINS_COLUMNS.disabled],
+            timer: result[DOMAINS_COLUMNS.timer],
+          }
+          response.status(200).send(resultResponse)
+        } else {
+          respondWithError(
```
Comment (alecps) [inline packages/phone-number-privacy/signer/src/domain/domain.service.ts]: We should probably tell the client that the domain does not exist in the db rather than return a 500.. Feel free to add a new error type in the common package for this

File: packages/phone-number-privacy/signer/src/domain/domain.service.ts [hunk 20]
```diff
@@ -1,49 +1,75 @@
...
 
-    const disabled = request.body.value ?? true
-    const result = await setDomainDisabled(request.params.domain, disabled, logger)
-
-    if (result) {
-      response.sendStatus(200)
-    } else {
-      response.status(500).send({ message: 'Cannot disable the domain' })
-    }
+    return setDomainDisabled(request.body.domain, logger)
```
Comment (alecps) [inline packages/phone-number-privacy/signer/src/domain/domain.service.ts]: We should first check that the domain exists in the db and is not already disabled. We can return different messages for both of these cases

**Review by alecps — COMMENTED** [review]
Love the fast response 🔥 Left a couple more minor comments

File: packages/phone-number-privacy/common/src/interfaces/requests.ts [hunk 21]
```diff
@@ -20,4 +20,8 @@ export interface GetQuotaRequest {
   sessionID?: string
 }
 
+export interface DomainRequestBody {
+  domain: string
+}
+
```
Comment (nategraf) [inline packages/phone-number-privacy/common/src/interfaces/requests.ts]: It's unclear what this request it? A request to get a domain? Could you add a comment or rename this to make it more clear.

File: packages/phone-number-privacy/common/src/interfaces/responses.ts [hunk 22]
```diff
@@ -22,3 +22,10 @@ export interface GetQuotaResponse {
   performedQueryCount: number
   totalQuota: number
 }
+
+export interface DomainStatusResponse {
+  domain: string
+  counter: number
+  disabled: boolean
+  timer: number
+}
```
Comment (nategraf) [inline packages/phone-number-privacy/common/src/interfaces/responses.ts:31]: This interface needs to be more general. Although the `SequentialDelayDomain` is the first we are implementing, it is not the only one. https://github.com/celo-org/celo-proposals/blob/master/CIPs/cip-0040.md#quota-status

File: packages/phone-number-privacy/common/src/interfaces/responses.ts [hunk 23]
```diff
@@ -22,3 +22,10 @@ export interface GetQuotaResponse {
   performedQueryCount: number
   totalQuota: number
 }
+
+export interface DomainStatusResponse {
+  domain: string
```
Comment (nategraf) [inline packages/phone-number-privacy/common/src/interfaces/responses.ts:27]: So although Domains will be sent across the wire JSON encoded, we should handle them in their typed object form whenever possible. I currently have the type definitions mixed into my domain hashing PR, but I'd be happy to split them out so this work can be based on it https://github.com/celo-org/celo-monorepo/pull/8705/files#diff-f19f568c47c9c7cf57b2cc5ce3c789dee075a3da138cf6a454e3ca5e6fdf615a

File: packages/phone-number-privacy/signer/src/database/models/domain.ts [hunk 24]
```diff
@@ -0,0 +1,20 @@
+export const DOMAINS_TABLE = 'domains'
+export enum DOMAINS_COLUMNS {
+  domain = 'domain',
```
Comment (nategraf) [inline packages/phone-number-privacy/signer/src/database/models/domain.ts]: If this is used as a lookup key, we should use the hash, as that is the form which is defined to be fully deterministic (i.e. the semantically same domain will always map to the same hash) as opposed to, for example, the JSON form which does not have a canonical serialization. Additionally, the hash is short and pseudorandom, making it a better key for lookups. We also don't have to store the full domain, although we can chose to do so, because the client will send it with each request.

File: packages/phone-number-privacy/signer/src/database/models/domain.ts [hunk 25]
```diff
@@ -0,0 +1,20 @@
...
+  counter = 'counter',
+  timer = 'timer',
+  disabled = 'disabled',
```
Comment (nategraf) [inline packages/phone-number-privacy/signer/src/database/models/domain.ts]: As mentioned above, this needs to support the addition of new domains in the future with unique state representations.

File: packages/phone-number-privacy/signer/src/database/wrappers/domain.ts [hunk 26]
```diff
@@ -0,0 +1,77 @@
...
+      .timeout(DB_TIMEOUT)
+    disableDomainMeter()
+    return true
+  } catch (err) {
+    Counters.databaseErrors.labels(Labels.update).inc()
+    logger.error(ErrorMessage.DATABASE_UPDATE_FAILURE)
+    logger.error(err)
+    disableDomainMeter()
+    throw err
+  }
```
Comment (nategraf) [inline packages/phone-number-privacy/signer/src/database/wrappers/domain.ts]: Just a simple suggestion to use the `finally` clause.
```suggestion
    return true
  } catch (err) {
    Counters.databaseErrors.labels(Labels.update).inc()
    logger.error(ErrorMessage.DATABASE_UPDATE_FAILURE)
    logger.error(err)
    throw err
  } finally {
    disableDomainMeter()
  }
```

File: packages/phone-number-privacy/signer/src/database/models/domain.ts [hunk 27]
```diff
@@ -0,0 +1,20 @@
...
+}
+export class Domain {
```
Comment (nategraf) [inline packages/phone-number-privacy/signer/src/database/models/domain.ts]: I'd recommend renaming this to `DomainState`, and renaming related function along with it, as this is distinct from the `Domain` itself.

File: packages/phone-number-privacy/common/src/interfaces/requests.ts [hunk 28]
```diff
@@ -20,4 +20,8 @@ export interface GetQuotaRequest {
   sessionID?: string
 }
 
+export interface DomainRequestBody {
```
Comment (nategraf) [inline packages/phone-number-privacy/common/src/interfaces/requests.ts]: I see this is used to implement the quota status and disable domain endpoints, but it does not currently include the domain options which is required to handle authentication. (e.g. the `SequentialDelayDomain` needs to require a signature on requests to disable the domain to prevent DoS attacks) https://github.com/celo-org/celo-proposals/blob/master/CIPs/cip-0040.md#quota-status

File: packages/phone-number-privacy/signer/src/domain/domain.service.ts [hunk 29]
```diff
@@ -0,0 +1,75 @@
...
+} from '@celo/phone-number-privacy-common'
+import { Request, Response } from 'express'
+import { respondWithError } from '../common/error-utils'
+import { DOMAINS_COLUMNS } from '../database/models/domain'
+import { getDomain, setDomainDisabled } from '../database/wrappers/domain'
+import { Endpoints } from '../server'
+import { IDomainService } from './domain.interface'
+
+export class DomainService implements IDomainService {
+  public async handleDisableDomain(
```
Comment (nategraf) [inline packages/phone-number-privacy/signer/src/domain/domain.service.ts:15]: We will need to add domain-specific authentication (e.g. signatures) here before moving this to production.

File: packages/phone-number-privacy/signer/src/domain/domain.service.ts (hunk 19 above)
Comment (nategraf) [reply packages/phone-number-privacy/signer/src/domain/domain.service.ts]: Additionally, asking about the status of a domain that this service does not know about is not an error. Instead it should return the initial state (e.g. for the `SequentialDelayDomain` this should be either an empty object or a status object with `counter` and `timer` set to `0`)

File: packages/phone-number-privacy/signer/src/server.ts [hunk 30]
```diff
@@ -17,8 +18,14 @@ export enum Endpoints {
   METRICS = '/metrics',
   GET_BLINDED_MESSAGE_PARTIAL_SIG = '/getBlindedMessagePartialSig',
   GET_QUOTA = '/getQuota',
+
+  DISABLE_DOMAIN = '/domain/disable',
+  DOMAIN_SIGN = '/domain/sign/',
+  DOMAIN_STATUS = '/domain/status',
```
Comment (nategraf) [inline packages/phone-number-privacy/signer/src/server.ts]: Overall, I like these paths better than what I suggested in CIP-40. I'll go ahead and amend that CIP to match this. A would suggest `quotaStatus` instead of `status` to clarify that this is not the status endpoint for the service itself.
```suggestion
  DOMAIN_STATUS = '/domain/quotaStatus',
```

File: packages/phone-number-privacy/signer/src/server.ts [hunk 31]
```diff
@@ -36,15 +43,21 @@ export function createServer() {
 
   const addMeteredEndpoint = (
     endpoint: Endpoints,
-    handler: (req: Request, res: Response) => Promise<void>
+    handler: (req: Request, res: Response) => Promise<void>,
+    method: 'post' | 'get' = 'post'
```
Comment (nategraf) [inline packages/phone-number-privacy/signer/src/server.ts:47]: I don't think there are any `GET` endpoints defined because all the new requests use the body to send their parameters. Why is handling of `GET` added here now?

**Review by nategraf — COMMENTED** [review]
Some comments from my first review. This isn't really a complete implementation of these methods as designed in [CIP-40](https://github.com/celo-org/celo-proposals/blob/master/CIPs/cip-0040.md) because it does not handle new domains in the future or options for authentication.

**Comment by nategraf** [comment]
I went ahead and opened up https://github.com/celo-org/celo-proposals/pull/289 to batch any changes to the spec in CIP-40 that need to be made if we make improvements during implementation. I haven't gone through a thorough review to determine any inconsistencies between the spec and this PR yet, but any that you know of, please either make a note so that we can update the spec, or change the implementation to match.

File: packages/phone-number-privacy/signer/src/domain/domain.service.ts (hunk 19 above)
Comment (alecps) [reply packages/phone-number-privacy/signer/src/domain/domain.service.ts]: agreed

File: packages/phone-number-privacy/signer/src/server.ts (hunk 30 above)
Comment (alecps) [reply packages/phone-number-privacy/signer/src/server.ts]: +1

File: packages/phone-number-privacy/signer/src/server.ts (hunk 31 above)
Comment (alecps) [reply packages/phone-number-privacy/signer/src/server.ts:47]: I think just for forwards compatibility. An earlier version of this PR used GET for the status endpoint and it doesn't seem that there's any harm in keeping this.

File: packages/phone-number-privacy/signer/src/domain/domain.service.ts (hunk 29 above)
Comment (alecps) [reply packages/phone-number-privacy/signer/src/domain/domain.service.ts:15]: Agreed. Let's do that in a separate PR / ticket

File: packages/phone-number-privacy/signer/src/database/models/domain.ts (hunk 25 above)
Comment (alecps) [reply packages/phone-number-privacy/signer/src/database/models/domain.ts]: I'm not sure that it does. 
- all domains will require the 'disabled' column
- Many domains will have some timestamp that they store. 
- Many domains will have a counter to track how many requests have been made
- New columns can be added as needed 
- Domains do not need to use every column in the table

File: packages/phone-number-privacy/signer/src/database/models/domain.ts (hunk 24 above)
Comment (alecps) [reply packages/phone-number-privacy/signer/src/database/models/domain.ts]: Yea we intend to store the hash. Is the domain serialization work at all relevant to how we hash the domain?

File: packages/phone-number-privacy/common/src/interfaces/responses.ts (hunk 23 above)
Comment (alecps) [reply packages/phone-number-privacy/common/src/interfaces/responses.ts:27]: I personally feel it would be easier to manage this in multiple PRs

File: packages/phone-number-privacy/common/src/interfaces/responses.ts (hunk 22 above)
Comment (alecps) [reply packages/phone-number-privacy/common/src/interfaces/responses.ts:31]: Agreed. We will add support for different domain types in a future PR

File: packages/phone-number-privacy/common/src/interfaces/requests.ts (hunk 21 above)
Comment (alecps) [reply packages/phone-number-privacy/common/src/interfaces/requests.ts]: Agreed. @emilhotkowski Can you please rename + revise this to match DomainRestrictedSignatureRequest type defined in CIP-40 https://github.com/celo-org/celo-proposals/blob/master/CIPs/cip-0040.md#extension-for-domain-restricted-requests

File: packages/phone-number-privacy/common/src/interfaces/requests.ts (hunk 28 above)
Comment (alecps) [reply packages/phone-number-privacy/common/src/interfaces/requests.ts]: The domain options are of type EIP712Object. Should this not wait until the serialization work is finished?

**Comment by alecps** [comment]
In general, this PR is not intended to be a finished implementation of these endpoints. Future tickets/PRs for versioning/code-refactor, domain options, authentication etc. will complete the implementation according to the cip40 spec. This approach is intended to make code review / management easier.

File: packages/phone-number-privacy/signer/src/database/models/domain.ts (hunk 25 above)
Comment (nategraf) [reply packages/phone-number-privacy/signer/src/database/models/domain.ts]: I will note that disabling a domain is not a required feature of all domains, and unauthenticated domains will most likely not implement it as there would be no way to prevent DoS attacks.

I don't share the intuition that these fields will be shared among many domains, but also don't see this as a major issue. It's true that we can implement new domains in a shared table where each uses a subset of the columns, and there is no need to change this approach unless things get out of hand. If that happens, we would need a database migration, but that doesn't seem like a major issue.

TL;DR happy to defer to y'all's opinion on this.

File: packages/phone-number-privacy/signer/src/database/models/domain.ts (hunk 24 above)
Comment (nategraf) [reply packages/phone-number-privacy/signer/src/database/models/domain.ts]: Ya, when I say serialization, its possibly more accurate to say hashing. (EIP-712 is kind of both).

File: packages/phone-number-privacy/common/src/interfaces/responses.ts (hunk 23 above)
Comment (nategraf) [reply packages/phone-number-privacy/common/src/interfaces/responses.ts:27]: Sounds good to me 👍

File: packages/phone-number-privacy/signer/src/database/models/domain.ts (hunk 25 above)
Comment (alecps) [reply packages/phone-number-privacy/signer/src/database/models/domain.ts]: Ahh I misunderstood the CIP wrt disabling domains. I agree we can keep this as is for simplicity now and do a db migration when / if that comes up. Good thing to have on our radar for sure though

**Review by alecps — APPROVED** [review]
Looks great! 🚀
